sql: prepare support for INSERT, UPDATE, DELETE

madelynnblue opened the pull request:
Make sure prepare statements are run in a transaction.

Use a new boolean flag on the planner to indicate prepare. Statement
implementations will gain knowledge of work to omit during prepare. An
alternative implementation would split up the prepare and exec portions of
the work. However, this would require passing around significant amounts
of state and I think the best implementation is the one here.

Keep a whitelist of statement types that support preparation instead of
attempting to exec them during prepare.

This also functions as a refactor for prepare statements in
general. Previously, we called parser.InferArgs, which walked the statement
and ran TypeCheck on all expressions. It turns out this is not needed,
since Select (and others) do this themselves. Thus, we can invoke those
worker functions directly but teach them about prepareOnly so they can
omit work once the val args have been populated and the type checks done.

Fixes #3819

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4037)

<!-- Reviewable:end -->

petermattis commented on sql/pgwire_test.go:
```diff
@@ -316,6 +313,103 @@ func TestPGPrepared(t *testing.T) {
+		{
+			"CREATE TABLE d.t (i INT, s STRING, d INT)",
+			nil,
+		},
```
Sort of strange that these are test cases. They are really part of test setup. I think it would be cleaner to pull them out and actually perform them as separate setup steps. You can do something like:

```
  setup := `CREATE DATABASE d; CREATE TABLE d.t (i INT, s STRING, d INT);`
  if err := db.Exec(setup); err != nil {
    t.Fatal(err)
  }
```

petermattis commented on sql/pgwire_test.go:
```diff
@@ -316,6 +313,103 @@ func TestPGPrepared(t *testing.T) {
+
+	testFailures := map[string]string{
+		"CREATE TABLE $1 (id INT)":  "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
+		"DROP TABLE t":              "pq: prepare statement not supported: *parser.DropTable",
```
Can we change this error so that it references `DROP TABLE` instead of `*parser.DropTable`? We should try to avoid leaking implementation details like the names of our internal structures into the end-user error messages.

petermattis commented on sql/plan.go:
```diff
@@ -152,6 +156,25 @@ func (p *planner) makePlan(stmt parser.Statement, autoCommit bool) (planNode, *r
+	case *parser.Update:
+		return p.Update(n)
+	default:
+		return nil, roachpb.NewUErrorf("prepare statement not supported: %T", stmt)
```
Perhaps `stmt.PGTag()` (whenever that PR goes in).

petermattis commented on sql/update.go:
```diff
@@ -132,6 +132,55 @@ func (p *planner) Update(n *parser.Update) (planNode, *roachpb.Error) {
+		}
+
+		return nil, nil
+	}
```
Some of our methods are becoming gigantic!

madelynnblue committed:
sql: prepare support for INSERT, UPDATE, DELETE

Make sure prepare statements are run in a transaction.

Use a new boolean flag on the planner to indicate prepare. Statement
implementations will gain knowledge of work to omit during prepare. An
alternative implementation would split up the prepare and exec portions of
the work. However, this would require passing around significant amounts
of state and I think the best implementation is the one here.

Keep a whitelist of statement types that support preparation instead of
attempting to exec them during prepare.

This also functions as a refactor for prepare statements in
general. Previously, we called parser.InferArgs, which walked the statement
and ran TypeCheck on all expressions. It turns out this is not needed,
since Select (and others) do this themselves. Thus, we can invoke those
worker functions directly but teach them about prepareOnly so they can
omit work once the val args have been populated and the type checks done.

Fixes #3819

mjibson commented:
Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/pgwire_test.go, line 330 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4037#-K9DrXjZBwSknQiM1aVI-r1-330:-K9EVoXTToVnqbggxLNX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/998f56bb70274dcb0f765b646b1c97eac65f98f2/sql/pgwire_test.go#L330)):*
Done.

---

*[sql/pgwire_test.go, line 396 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4037#-K9DrXjZBwSknQiM1aVI-r1-396:-K9EUwCZUNtQZonMyPWJ:-625985944) ([raw file](https://github.com/cockroachdb/cockroach/blob/998f56bb70274dcb0f765b646b1c97eac65f98f2/sql/pgwire_test.go#L396)):*
Adding a TODO for this, discussed below.

---

*[sql/plan.go, line 171 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4037#-K9DrXjZBwSknQiM1aVJ-r1-171:-K9EVGvDuI9aur5g6Qqj:-1018314897) ([raw file](https://github.com/cockroachdb/cockroach/blob/998f56bb70274dcb0f765b646b1c97eac65f98f2/sql/plan.go#L171)):*
Adding a TODO for that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4037)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4037#-:-K9FGwMSaeG6pAowVPfV:676834233)*
<!-- Sent from Reviewable.io -->
